File: Task 1/Java/app/src/main/java/dev/ogabek/java/activity/MainActivity.java
package dev.ogabek.java.activity;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.content.DialogInterface;
import android.os.AsyncTask;
import android.os.Bundle;
import android.view.View;
import android.widget.EditText;

import com.google.android.material.floatingactionbutton.FloatingActionButton;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutionException;

import dev.ogabek.java.R;
import dev.ogabek.java.adapter.NoteAdapter;
import dev.ogabek.java.database.NoteRepository;
import dev.ogabek.java.model.Note;

public class MainActivity extends AppCompatActivity {

    private RecyclerView rv_notes;
    private FloatingActionButton btn_add;

    private List<Note> notes = new ArrayList<Note>();
    private NoteAdapter adapter = new NoteAdapter(this, notes);

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        NoteRepository repository = new NoteRepository(getApplication());

        try {
            notes.addAll(new GetNotes(repository).execute().get());
        } catch (ExecutionException | InterruptedException e) {
            e.printStackTrace();
        }

        initViews();

    }

    private void initViews() {

        btn_add = findViewById(R.id.btn_add);
        rv_notes = findViewById(R.id.rv_notes);

        btn_add.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View ui) {

                View view = getLayoutInflater().inflate(R.layout.custom_dialog, null);
                AlertDialog alertDialog = new AlertDialog.Builder(MainActivity.this).create();
                alertDialog.setTitle("New Note");
                alertDialog.setCancelable(true);

                EditText text = view.findViewById(R.id.etComments);

                alertDialog.setButton(AlertDialog.BUTTON_POSITIVE, "Save", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int i) {
                        NoteRepository repository = new NoteRepository(getApplication());
                        int id = 0;
                        try {
                            id = new GetNotes(repository).execute().get().size();
                        } catch (ExecutionException | InterruptedException e) {
                            e.printStackTrace();
                        }

                        Note note = new Note(id, text.getText().toString());
                        new SaveNote(repository).execute(note);
                        notes.add(note);
                        adapter.notifyDataSetChanged();
                    }
                });

                alertDialog.setButton(AlertDialog.BUTTON_NEGATIVE, "Cancel", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialogInterface, int i) {
                        alertDialog.cancel();
                    }
                });

                alertDialog.setView(view);
                alertDialog.show();

            }
        });

        rv_notes.setLayoutManager(new GridLayoutManager(this, 1));
        rv_notes.setAdapter(adapter);

    }


    // Create new Thread for saving note
    class SaveNote extends AsyncTask<Note, Void, Void> {
        NoteRepository repository;

        SaveNote(NoteRepository repository) {
            this.repository = repository;
        }

        @Override
        protected Void doInBackground(Note... notes) {
            repository.saveNote(notes[0]);
            return null;
        }
    }

    // Create new Thread for getting Notes
    class GetNotes extends AsyncTask<Void, Void, List<Note>> {
        NoteRepository repository;

        GetNotes(NoteRepository repository) {
            this.repository = repository;
        }

        @Override
        protected List<Note> doInBackground(Void... unused) {
            return repository.getNotes();
        }

        @Override
        protected void onPostExecute(List<Note> users) {
            super.onPostExecute(users);
        }
    }

    // Create new Thread for clearing Notes
    class ClearNotes extends AsyncTask<Void, Void, Void> {
        NoteRepository repository;

        ClearNotes(NoteRepository repository) {
            this.repository = repository;
        }

        @Override
        protected Void doInBackground(Void... unused) {
            repository.clearData();
            return null;
        }

        @Override
        protected void onPostExecute(Void unused) {
            super.onPostExecute(unused);
        }
    }

}

File: Task 1/Kotlin/app/src/main/java/dev/ogabek/kotlin/activity/MainActivity.kt
package dev.ogabek.kotlin.activity

import android.content.ContentValues.TAG
import android.content.DialogInterface
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.View
import android.widget.EditText
import androidx.appcompat.app.AlertDialog
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.floatingactionbutton.FloatingActionButton
import dev.ogabek.kotlin.R
import dev.ogabek.kotlin.adapter.NoteAdapter
import dev.ogabek.kotlin.database.NoteRepository
import dev.ogabek.kotlin.model.Note
import java.util.concurrent.ExecutionException
import java.util.concurrent.Executors
import kotlin.math.log

class MainActivity : AppCompatActivity() {

    private lateinit var rv_notes: RecyclerView
    private lateinit var btn_add: FloatingActionButton

    private val notes: ArrayList<Note> = ArrayList()
    private lateinit var adapter: NoteAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        initViews()


    }

    private fun initViews() {

        val repository = NoteRepository(application)

        notes.addAll(getAllNotes(repository))

        adapter = NoteAdapter(this, notes)
        rv_notes = findViewById(R.id.rv_notes)
        rv_notes.layoutManager = LinearLayoutManager(this)
        rv_notes.adapter = adapter

        btn_add = findViewById(R.id.btn_add)

        btn_add.setOnClickListener {
            val view = layoutInflater.inflate(R.layout.custom_dialog, null)
            val alertDialog = AlertDialog.Builder(this).create()
            alertDialog.setTitle("New Note")
            alertDialog.setCancelable(true)

            val text = view.findViewById<EditText>(R.id.etComments)

            alertDialog.setButton(AlertDialog.BUTTON_POSITIVE, "Save") { _, _ ->
                val repository = NoteRepository(application)
                val id = notes.size
                val note = Note(id, text.text.toString())
                saveNote(repository, note)
                notes.add(note)
                adapter.notifyDataSetChanged()
            }

            alertDialog.setButton(AlertDialog.BUTTON_NEGATIVE, "Cancel") { _, _ ->
                alertDialog.cancel()
            }

            alertDialog.setView(view)
            alertDialog.show()
        }

    }

    private fun saveNote(repository: NoteRepository, note: Note) {
        val executor = Executors.newSingleThreadExecutor()  // in Background
        executor.execute {
            repository.saveNote(note)
        }
    }

    private fun getAllNotes(repository: NoteRepository): List<Note> {
        Log.d(TAG, "getAllNotes: ${repository.getNotes()}")
        return repository.getNotes()
    }

    fun deleteData(repository: NoteRepository) {
        val executor = Executors.newSingleThreadExecutor()  // in Background
        executor.execute {
            repository.deleteData()
        }
    }

}